### PR TITLE
documents: optimize serialization

### PR DESCRIPTION
**Pull request by Garfield-fr**
Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## How to test?

- Check organisation/library facet on all application.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/api.py [hunk 1]
```diff
@@ -90,6 +90,16 @@ def flush_and_refresh(cls):
...
         current_search.flush_and_refresh(cls.Meta.index)
 
+    def get_record_by_pid(self, pid, fields=None):
+        """Search by pid."""
+        query = self.filter('term', pid=pid).extra(size=1)
+        if fields:
+            query = query.source(includes=fields)
+        response = query.execute()
+        if response.hits.total.value != 1:
+            raise Exception('Result not found.')
```
Comment (rerowep) [inline rero_ils/modules/api.py:100]: please change to `raise ValueError(f'Record not found pid: {pid})`

File: rero_ils/modules/libraries/api.py [hunk 2]
```diff
@@ -63,6 +63,15 @@ class Meta():
 
         default_filter = None
 
+    def get_libraries_by_organisation_pid(self, org_pid, fields=None):
+        """Search libraries by organisation pid."""
+        query = self.filter('term', organisation__pid=org_pid)
+        if (fields):
```
Comment (rerowep) [inline rero_ils/modules/libraries/api.py:69]: No parentheses needed here.

File: rero_ils/modules/organisations/api.py [hunk 3]
```diff
@@ -54,6 +54,17 @@ class Meta:
...
+    def get_record_by_viewcode(self, viewcode, fields=None):
+        """Search by viewcode."""
+        query = self.filter('term', code=viewcode).extra(size=1)
+        if fields:
+            query = query.source(includes=fields)
+        response = query.execute()
+        if response.hits.total.value != 1:
+            raise Exception(
+                f'Organisation viewcode {viewcode}: Result not found.')
+        return response.hits.hits[0]._source
```
Comment (rerowep) [inline rero_ils/modules/organisations/api.py:66]: Is `_source` the same as `to_dict()`?

**Review by rerowep — CHANGES_REQUESTED** [review]

File: rero_ils/modules/organisations/api.py (hunk 3 above)
Comment (Garfield-fr) [reply rero_ils/modules/organisations/api.py:66]: No, it's a record key on elasticsearch result

File: rero_ils/modules/api.py [hunk 4]
```diff
@@ -90,6 +90,16 @@ def flush_and_refresh(cls):
...
         current_search.flush_and_refresh(cls.Meta.index)
 
+    def get_record_by_pid(self, pid, fields=None):
+        """Search by pid."""
+        query = self.filter('term', pid=pid).extra(size=1)
+        if fields:
+            query = query.source(includes=fields)
+        response = query.execute()
+        if response.hits.total.value != 1:
+            raise ValueError(f'Record not found pid: {pid}')
```
Comment (zannkukai) [inline rero_ils/modules/api.py:100]: Maybe `NotFound` build-in error could be a better Exception option ?

File: rero_ils/modules/organisations/api.py [hunk 5]
```diff
@@ -54,6 +54,17 @@ class Meta:
...
         default_filter = None
 
+    def get_record_by_viewcode(self, viewcode, fields=None):
+        """Search by viewcode."""
+        query = self.filter('term', code=viewcode).extra(size=1)
+        if fields:
+            query = query.source(includes=fields)
+        response = query.execute()
+        if response.hits.total.value != 1:
+            raise Exception(
```
Comment (zannkukai) [inline rero_ils/modules/organisations/api.py:64]: same here

**Review by zannkukai — APPROVED** [review]

**Review by rerowep — APPROVED** [review]

**Review by sebastiendeleze — APPROVED** [review]

File: rero_ils/modules/api.py [hunk 6]
```diff
@@ -90,6 +90,16 @@ def flush_and_refresh(cls):
         """Flush and refresh index."""
         current_search.flush_and_refresh(cls.Meta.index)
 
+    def get_record_by_pid(self, pid, fields=None):
+        """Search by pid."""
+        query = self.filter('term', pid=pid).extra(size=1)
+        if fields:
+            query = query.source(includes=fields)
+        response = query.execute()
+        if response.hits.total.value != 1:
```
Comment (iGormilhit) [inline rero_ils/modules/api.py:99]: What if they are two or more hits? Should the message be something like " :scream: "?

**Review by iGormilhit — COMMENTED** [review]
Could the commit message be more informative?

File: rero_ils/modules/api.py (hunk 6 above)
Comment (Garfield-fr) [reply rero_ils/modules/api.py:99]: A pid value is unique. Either it exists or you have an exception?

**Review by iGormilhit — APPROVED** [review]

**Pull request merged**